### DISCUSSION TITLE
Web interface discussion

### DISCUSSION THREAD

**Discussion by albar965** [albar965/littlenavmap#1152] [General]
I don't know if it make sense to add this to one of the numerous PRs. Therefore, I'll just start a discussion here.

@u-an-i Can we rename the names in the dropdown box in the web interface to make them more user friendly:

1. "ol-map" -> "OpenLayers Dynamic Map"
2. "no exclusive plugin" -> "Static Map"

A lot of people use the web interface and most are not developers. :wink:
The names are just suggestions. Let me know if you have a better idea.

Also, I think the legacy map can be removed for good. No need for three map styles.

@KOKAProduktion or @u-an-i : Can we be sure that the new address `http://MYCOMPUTER:8965/html/2021-a/index.html` does not break these addons: https://flightsimulator.me/simulators/flight-simulator/panels/105-ingamepanel-navmap-for-flight-simulator and https://flightsim.to/file/43086/little-navmap-vr-panel ? 

Not sure. Maybe the API is not affected, is it?

Sorry for being a pedantic nuisance. :upside_down_face: 

Alex

**Comment by u-an-i** [comment]
@albar965 no problem,

the name of the folder of the plugin is the name shown in the menu. I don‘t know whether space and uppercase work. I‘m currently out of reach of my development machine. If they don‘t, i can think of a mean to seperate the name of the plugin from its path.

web api is solely server-delivered. The html files don‘t matter except someone lets fetch one of them, then their path would have to be adjusted.

The url is sth which can be cosmetically fixed to only be the server domain.


Legacy: if you like we can remove the toggle from the menu. I would leave the files so that someone can use them if desired and see how to work without web api.

The folder structure itself was prework currently not being necessary: it allows „old versions“ of the web frontend to be accessible for those who grew used to a particular one in case someone makes breaking changes in a new one.

**Reply by u-an-i** [reply]
@bkielczewski you are the author of the vr panel. Is it affected by the path changes?

**Reply by albar965** [reply]
> the name of the folder of the plugin is the name shown in the menu. I don‘t know whether space and uppercase work. I‘m currently out of reach of my development machine. If they don‘t, i can think of a mean to seperate the name of the plugin from its path.

Perfect would be to put the name into a JSON plugin description.
You can also hardcode the names in the API as a workaround. Good enough.

> Legacy: if you like we can remove the toggle from the menu. I would leave the files so that someone can use them if desired and see how to work without web api.

Perfect solution for me. 

Alex

**Reply by u-an-i** [reply]
JSON plugin description :  good idea! Will do as soon as possible.

Will remove the menu entry then too.

**Reply by albar965** [reply]
Thanks. I intend to make a public 3.0.7.rc1 the next days or week. I have to fly over LNM a few times (aka testing :upside_down_face: ) before.

**Reply by bkielczewski** [reply]
@u-an-i yes, I'm an author of [the panel](https://flightsim.to/file/43086/little-navmap-vr-panel). I'm not affected by anything inside `web` you have, since I replace that with my web content. I would be affected if document root will be served from `/html` but I can adapt to this.

**Reply by u-an-i** [reply]
@albar965 i plan to overhaul the plugin and toolbar interface and make it completely JSON configuration based. The toolbar API’s functions took a JSON configuration already only for example, those would be moved from a code call to the static JSON file being the plugin configuration. The `init` function call will be replaced by the use of JS’/Browser‘s `postMessage`. This is slightly uncomfortable to use for the developer and the success of the cleanup function call i cannot catch without guessing it should have suceeded after x milliseconds but it completes security, i will make it as easy to use as possible and automatic to understand from the documentation as i achieve.

**Reply by albar965** [reply]
> @albar965 i plan to overhaul the plugin and toolbar interface and make it completely JSON configuration based. The toolbar API’s functions took a JSON configuration already only for example, those would be moved from a code call to the static JSON file being the plugin configuration. The `init` function call will be replaced by the use of JS’/Browser‘s `postMessage`. This is slightly uncomfortable to use for the developer and the success of the cleanup function call i cannot catch without guessing it should have suceeded after x milliseconds but it completes security, i will make it as easy to use as possible and automatic to understand from the documentation as i achieve.

Sounds good. I continue with 3.0.7.rc1 public (not sure when). Your change can also be added to the next rc or the release. No problem.

Alex

**Comment by bkielczewski** [comment]
Ok, just fyi, here's what I'm using at the moment:

- `http://localhost:8965` <- assuming document root is served from here
- Not relevant here probably, but I'm parsing the output of tags: `{airportText}`, `{airportRunwayText}`, `{airportComText}`, `{airportProcedureText}`, `{airportWeatherText}`, `{flightplanText}`, `{aircraftProgressText}`, and I depend on what's in there since don't have an API for that.
- I'm calling `/mapimage` to get map tiles
- From API I'm calling `'/api/sim/info'`, `/api/map/features`

If you were to move document root from `/` to `/html`, I could check for `/html` first, and if that would return 404 I could fall back to `/`,

**Reply by u-an-i** [reply]
Document root remains `/`

**Comment by albar965** [comment]
Anything to add or correct in the changelog?

https://github.com/albar965/littlenavmap/blob/3203169ed8fcde61538fa6e70267c9ba6cf9c78b/CHANGELOG.txt#L56-L70
and
https://github.com/albar965/littlenavmap/blob/3203169ed8fcde61538fa6e70267c9ba6cf9c78b/CHANGELOG.txt#L161-L175

Alex

**Reply by u-an-i** [reply]
good!
  
hint: use "random flight" instead of "random plan" throughout 1. . 1.'s last point typo: "destionation"  
  
2.'s last point: "web frontend in general, web API, web plugin API, web toolbar API and web theming CSS templates; useful for developers and people who wish to enhance or modify the web frontend experience." instead of "web API".

**Reply by albar965** [reply]
Thanks. Done.

**Comment by albar965** [comment]
@u-an-i  and @KOKAProduktion 
I've got this report from a user where the log ends with before crashing:

```
[2024-07-10 7:47:20.425 DEBUG]: TemplateCache: trying cached /progress_doc
[2024-07-10 7:47:20.425 DEBUG]: TemplateCache: trying cached /progress_doc
[2024-07-10 7:47:20.426 WARN ]: Trying to construct an instance of an invalid type, type id: 206946128
[2024-07-10 7:47:20.426 WARN ]: Trying to construct an instance of an invalid type, type id: 206942544
[2024-07-10 7:47:20.426 WARN ]: Trying to construct an instance of an invalid type, type id: 206947728
```

Anybody has a clue where this happens?

I can also attach the full log if needed.

Alex

**Reply by bkielczewski** [reply]
Ignore if not relevant, but I'm still getting these crashes, now on 3.0.8 which don't really get logged. Only indication what happened is in windows event log. It looks it's an unhandled error that is causing the process to quit. Regardless of the actual cause for error, I would advise to look into handling these errors and try to log them.

That's a tail of my latest crash yesterday:

```
[2024-07-09 22:53:22.729 DEBUG]: QString RouteController::getFlightplanTableAsHtml(float, bool) const
[2024-07-09 22:53:25.556 DEBUG]: TemplateCache: trying cached /assets/lnm/progress
[2024-07-09 22:53:27.796 DEBUG]: HttpConnectionHandler (0x16e834f0): disconnected
[2024-07-09 22:53:30.631 DEBUG]: HttpConnectionHandler (0x16e834f0): handle new connection
[2024-07-09 22:53:30.631 DEBUG]: TemplateCache: trying cached /assets/lnm/progress
[2024-07-09 22:53:30.839 DEBUG]: HttpConnectionHandler (0x16e83f70): disconnected
[2024-07-09 22:53:32.775 DEBUG]: TemplateCache: trying cached /assets/lnm/flightplan
[2024-07-09 22:53:32.822 DEBUG]: QString RouteController::getFlightplanTableAsHtml(float, bool) const
[2024-07-09 22:53:35.194 DEBUG]: HttpConnectionHandler (0x16e84980): disconnected
[2024-07-09 22:53:35.714 DEBUG]: TemplateCache: trying cached /assets/lnm/progress
[2024-07-09 22:53:36.001 DEBUG]: HttpConnectionHandler (0x16e83f70): handle new connection
[2024-07-09 22:53:36.001 DEBUG]: HttpConnectionHandler (0x16e84980): handle new connection
[2024-07-09 22:53:37.978 DEBUG]: HttpConnectionHandler (0x16e834f0): disconnected
[2024-07-09 22:53:40.795 DEBUG]: HttpConnectionHandler (0x16e834f0): handle new connection
[2024-07-09 22:53:40.805 DEBUG]: TemplateCache: trying cached /assets/lnm/progress
[2024-07-09 22:53:40.925 DEBUG]: HttpConnectionHandler (0x16e83170): disconnected
[2024-07-09 22:53:42.871 DEBUG]: TemplateCache: trying cached /assets/lnm/flightplan
[2024-07-09 22:53:42.979 DEBUG]: QString RouteController::getFlightplanTableAsHtml(float, bool) const
[2024-07-09 22:53:45.230 DEBUG]: HttpConnectionHandler (0x1b864130): disconnected
[2024-07-09 22:53:45.814 DEBUG]: HttpConnectionHandler (0x16e83170): handle new connection
[2024-07-09 22:53:45.874 DEBUG]: TemplateCache: trying cached /assets/lnm/progress
[2024-07-09 22:53:46.027 DEBUG]: HttpConnectionHandler (0x1b864130): handle new connection
```

And that's related entry in windows event log:

```
Faulting application name: littlenavmap.exe, version: 0.0.0.0, time stamp: 0x66892697
Faulting module name: qsqlite.dll, version: 5.15.2.0, time stamp: 0x00000000
Exception code: 0xc0000005
Fault offset: 0x0000000000090ba4
Faulting process id: 0x0xE50
Faulting application start time: 0x0x1DAD24452329DDD
Faulting application path: C:\Users\bkiel\AppData\Local\Programs\Little Navmap\littlenavmap.exe
Faulting module path: C:\Users\bkiel\AppData\Local\Programs\Little Navmap\sqldrivers\qsqlite.dll
Report Id: ea038114-eeaf-482f-a137-12a5e4a7f6db
Faulting package full name: 
Faulting package-relative application ID: 
```

**Reply by albar965** [reply]
Maybe related to this: https://www.sqlite.org/threadsafe.html

All simulator databases are opened read-only but there can be still issues with internal SQLite structures.

**Reply by albar965** [reply]
QSqlite already opens the connection using ´SQLITE_OPEN_NOMUTEX`. 

https://github.com/qt/qtbase/blob/86c62c8f6088ec148512457cb7e964661ba643b0/src/plugins/sqldrivers/sqlite/qsql_sqlite.cpp#L774

So threading should be no issue.

**Reply by KOKAProduktion** [reply]
Hi guys!

Sorry to hear that... Sadly I don't have any idea where this might occur.

- Is there any user action related to the crash or is it crashing randomly? 
- Is it only happening when using the WebInterface? Does it occur without using the WebInterface?

I remember having to create another Marble or MarblePainter (?) instance back then for drawing tiles without features, so maybe it's a clue that I've messed up something there...

I will have a look into it at the weekend. Sorry I can't provide any help ATM.

Cheers!

**Reply by albar965** [reply]
There is another instance of MarbleWidget and all used query objects (MapQuery, etc.) duplicated for the web server.

The question is: Does the web server itself use these query classes from more than one thread without synchronization?
The query classes contain prepared statements which cannot be used multithreaded. The database can be used from multiple threads.

The only user action I've seen so far is using a MSFS plugin which relies on the web server.

**Reply by KOKAProduktion** [reply]
> Does the web server itself use these query classes from more than one thread without synchronization?

I don't recall that explicitly but I would assume that at least 2 "simultaneous" requests may create a condition like that, if the web server is providing a separate thread per request. I will check for that! Thanks!

**Reply by u-an-i** [reply]
Hi :(,

hi @KOKAProduktion , hi @albar965 , hi @bkielczewski . Like KOKA asked: do we know of a related user action?

The web frontend can only make LNM crash when a LNM service called from the web frontens crashes. Currently I'm aware of 3 services:
- map image delivery (user action: click and zoom, auto follow, center flight plan)
- web api (only used by Dynamic Touch and Drag Map in the ootb web frontend)
- HTML placeholder replacement (user action: open a tab containing such a file eg. Flight Plan)

I navigated through all HTML pages and used all features in the web frontend I thought of in RC 3.0.7 and just now in v.3.0.8. Just for the record I didn't experience a crash. I use the 64-bit LNM app.

The log snippets common entry is a `TemplateCache` and I see the url `/assets/lnm/progress` mentioned. I don't recognize this url. I don't remember a TemplateCache. That and `getFlightplanTableAsHtml` look like something might crash when doing the HTML placeholder replacement.

I found `\src\web\requesthandler.cpp:381`. @albar965 The HTML files location changed in the new release. Is perhaps a cache of an old location active? I skimmed through the code there: I didn't observe a suspicious action.

Regarding `Trying to construct an instance of an invalid type` this sounds like something related to `Q_DECLARE_METATYPE` or `QVariant` which I think I remember having come across in the web api code. Construction is a `new` operation. Such are to focus on when looking through the web api code.

And yes, Windows crash log looks like the crash occurs in conjunction with a SQLite access.

**Reply by bkielczewski** [reply]
Hey @u-an-i,

No user action, but the web server is being constantly molested by concurrent requests to these 3 services you've mentioned by my front-end. I imagine it would be hard to replicate by just human interaction.

`/assets/lnm/progress` it's from my custom web content and it just contains `{aircraftProgressText}`, basically same as `/progress_doc.html` in default web content.

Now thing is at some point in the past, on 3.0.6, I pinky-swear this crashed with an ordinary system dialog box showing SQL query error, but I blame myself I didn't screenshot it hoping it will show up again when I have time to do so. Funny thing is when few days ago I reverted from 3.0.7.rc1 to 3.0.6 it was stable for me. Then I tried 3.0.8 and have crashes again. Tried re-generating scenery database and re-installing Navigraph data, makes no difference.

**Reply by u-an-i** [reply]
@bkielczewski ah, thanks for clearing this up.

So the web server = LNM likely does receives so many requests that it might be working on some of them in parallel which might create a conflict in some shared global state which leads to code working on data not expected which is not shielded in the code and that data happens to crash the code = LNM.

The hint KOKA got to check prepared SQL statements appears possibly fruitful given Windows logged a crash in relation to the database DLL.

I don't currently know how the server code in LNM handles long-running tasks and whether it can work on requests in parallel so that I could suggest a way of serialising access with the prepared statements to the database.

**Reply by albar965** [reply]
> Now thing is at some point in the past, on 3.0.6, I pinky-swear this crashed with an ordinary system dialog box showing SQL query error, but I blame myself I didn't screenshot it hoping it will show up again when I have time to do so. Funny thing is when few days ago I reverted from 3.0.7.rc1 to 3.0.6 it was stable for me. Then I tried 3.0.8 and have crashes again.

There was a change in atools::sql::SqlDatabase which now passes the correct parameter down to open the databases now really readonly. Might be related. Can you see something in the log?

The problem is that I cannot reproduce any of these crashes here. Neither on Linux nor Windows. :slightly_frowning_face: 


> I don't currently know how the server code in LNM handles long-running tasks and whether it can work on requests in parallel so that I could suggest a way of serialising access with the prepared statements to the database.

I'd add a mutex around each call to MapQuery, AirportQuery and the like. I already had to do this for the MapWidget. Not the most elegant solution, I know.

And none of these query classes ever write to a simulator database which are opened read-only. Only userpoints, logbook and a few more are opened read-write but never scenery library databases.

**Reply by bkielczewski** [reply]
I think I can reliably reproduce it. Grab [jmeter](https://jmeter.apache.org/), and here's my plan: [LNM.zip](https://github.com/user-attachments/files/16169715/LNM.zip). You might want to change flightplan/progress requests to default web content equivalents.

Crash happens when there's an active sim connection, only takes 10 threads and few seconds. Without it, it can take much more punishment, would rather close connections then crash. I'll play around with it more, it's just too late now.

**Reply by u-an-i** [reply]
@bkielczewski 

https://github.com/user-attachments/assets/799eb827-0421-4368-b91d-43dc7d858893

I could **not** reproduce a crash with 32 threads, HTML paths adjusted to ootb paths and active sim connection in the current latest r/3.0 branch @albar965 using your plan.

With 33 threads, at least 1 thread would fail to retrieve a response. 

The summary report fail ratio is due to my trying with 100 threads before.

While the test was running, the app remained "responsive", the web server for the web browser did not.

I was using OpenStreet Maps theme.

Can you share your html files? Perhaps a certain replacement therein triggers the crash.

**Reply by u-an-i** [reply]
Sorry, sim connection did not establish using the built version.

3.0.8 crashed neither:


https://github.com/user-attachments/assets/1ea7c82e-959e-46fe-8a2f-82f0b2a380e2

**Reply by bkielczewski** [reply]
Thanks for checking. I need to spend time playing around with different combinations of things so maybe then I can give you a pattern you could reproduce. I wouldn't worry for just failed responses, the number of requests that makes with even few threads is too much to be realistic. it's just that crash that's bad.

**Reply by albar965** [reply]
Maybe it makes sense to track all usage of the MapQuery, AirportQuery and maybe other query classes and put a mutex around it.

In the meantime I found a [stacktrace reporter (cpptrace)](https://github.com/jeremy-rifkin/cpptrace) which actually works on all LNM supported platforms. This should help a bit to catch these issues.

BTW: I did not get further reports about similar crashes. Might happen rarely.

**Reply by bkielczewski** [reply]
Just to let you know, I've tried to reproduce it again with my plan (on 3.0.8) and failed this time, it just worked.
Also last few flights I didn't get crashes. So I'm like unable to give you anything at this time and that's a bit crazy. 
Anyway, I agree it's probably rare, I also didn't get reports like that from my panel users.
If you could add something that would make a log entry when the process quits that would be good for the future reports like that I imagine. If the crashes come back, I'll try to reproduce.

**Reply by albar965** [reply]
Thanks @bkielczewski .

I added more mutexes to database accesses in the webapi. This might help to get rid of the crashes no matter how rare.
There is more stuff which has to be synchronized like the user aircraft  which might change in the middle of an access.

**Comment by albar965** [comment]
@u-an-i  and @KOKAProduktion 

Sorry to be a PITA, but I see several issues with the web interface (no plugins activated, Firefox):
- Android (Samsung S23): The map fills only the top left quarter of the screen. Everything else working as expected. The issue disappeared after reloading. Probably no biggie.
- Lenovo P11: Map is fully visible, but tap to jump/move seems to jump erratically (i.e. to a nearby but a wrong position).
- Users report zooming not working on iPads which I cannot reproduce on a MacBook using Firefox.
- Starting on Safari I get a message "User interface not rendering correctly, switch to legacy". Which it does and this seems to work ok. I thought we removed the legacy. Better keep it for now.

The biggest issue I see so far is the click-to-move but maybe I do not understand the logic there.

Maybe it makes sense to always keep zoom and move buttons on the interface to workaround/mitigate such issues. Also keep legacy in the web interface menu. Sorry, I know it was my stupid idea to remove it.

I'll go ahead and make a release 3.0.9 anyway. Maybe we can improve the interface for 3.0.10.

Alex

**Reply by u-an-i** [reply]
No PITA, only scaring me I missed something. This is what I was working on, I will take a look once I‘m home. Always come forward when you experience something unexpected in the web frontend. I removed the switch to legacy, not legacy itself. In „current“ versions of Chrome, Safari and Firefox everything should work on Desktop and mobile without legacy.

**Reply by u-an-i** [reply]
@albar965 https://github.com/albar965/littlenavmap/pull/1177 fixes Firefox mobile / desktop touch / click move and corrects map size for iPad Safari.

Zooming works for me on my iPad, see video. Note: zoom is to current image center not pinched fingers center.

https://github.com/user-attachments/assets/187dad40-0182-41cc-90ce-b5a97e830746

(video recorded before iPad map size fix)


Safari Mac works for me in BrowserStack (real devices, real browsers):

![image](https://github.com/user-attachments/assets/ba5ac16d-51d6-4d17-ba50-6eb1ce8cac82)


The map not filling screen thing could have been due to old cached CSS, Safari (any "recent", but not on Windows!) all support the current web ui and iPad has the finger pinch zoom since I reworked the web ui in 2021. I hope reports regarding those browsers are only due to old versions.

**Reply by albar965** [reply]
Perfect! You fix things faster than I think. Good one for 3.0.10. :+1:

**Reply by albar965** [reply]
@u-an-i  and @KOKAProduktion :
I made the last two monstrous commits to release/3.0 because of an crash report sent my a user which indicated a synchronization problem when accessing the HtmlInfoBuilder ending in a crash in SQL classes using prepared statements. Easily found thanks to the stack trace now included. 

I cleaned the query class management where classes took queries from all over the place (NavApp). Always a bad idea and quite messy. HtmlInfoBuilder and MapPaintWidget and their aggregates now use `Queries` which bundles all like MapQuery and allows easy locking ([querymanager.h](https://github.com/albar965/littlenavmap/blob/release/3.0/src/query/querymanager.h).

I hope this helps with some of the crashes seen in the past.

Going further I noticed that you already use mechanisms to do the work in the main thread / eventqueue by using invoke (albeit with direct connection?) or queued connections. Not sure if all the sync stuff I added is needed but definitely for the HtmlInfoBuilder.

Are there any other places in web and webapi that might need synchronization?

**Reply by u-an-i** [reply]
web in LNM is the requesthandler. This basically still works like it used to. I am not aware of synchronisation needs therein. It defers its work to (a) MapPaintWidget if I remember correctly.

webapi @KOKAProduktion is expert on. I can have a look too if you're busy.

**Reply by albar965** [reply]
> web in LNM is the requesthandler. This basically still works like it used to. I am not aware of synchronisation needs therein. It defers its work to (a) MapPaintWidget if I remember correctly.

Thanks for confirmation. I guess there is no need to look closer. Most places I checked run in the main thread context.
